### PR TITLE
Get accurate statistic for zedstore

### PR DESCRIPTION
**Pull request by pengzhout**
The first two commit is for postgres upstream to extend related API.
The last one is for zedstore

### CODE REVIEW

File: src/backend/access/heap/heapam_handler.c [hunk 1]
```diff
@@ -977,10 +977,24 @@ heapam_relation_copy_for_cluster(Relation OldHeap, Relation NewHeap,
 	pfree(isnull);
 }
 
+static void
+heapam_scan_analyze_beginscan(Relation onerel, AcquireSampleContext *context)
+{
+	Assert(scan->rs_rd);
```
Comment (l-wang) [inline src/backend/access/heap/heapam_handler.c]: should be Assert(context->scan->rs_rd)

**Review by l-wang — CHANGES_REQUESTED** [review]
Hey @pengzhout , @tvesely and I tried to review it but we hit a SIGSEGV at initdb, could you please take a look at that and get back to us?

**Comment by pengzhout** [comment]
@l-wang @tvesely  My mistake, fixed one variable initialize issue, initdb works now, please take another look

**Comment by l-wang** [comment]
The server still crashes when I analyze a table, repro:
```
create table foo(a int) using zedstore;
insert into foo values (1);
analyze foo;
```

It looks like `context->extrainfo_slots[DISK_SIZE]` was free'd in `zedstoream_scan_analyze_endscan()` and then was accessed again in `compute_disk_stats()`

File: src/backend/commands/analyze.c [hunk 2]
```diff
@@ -83,24 +83,24 @@ int			default_statistics_target = 100;
 static MemoryContext anl_context = NULL;
 static BufferAccessStrategy vac_strategy;
 
-
 static void do_analyze_rel(Relation onerel,
 						   VacuumParams *params, List *va_cols,
 						   AcquireSampleRowsFunc acquirefunc, BlockNumber relpages,
 						   bool inh, bool in_outer_xact, int elevel);
+static void compute_disk_stats(VacAttrStats **stats, int natts,
```
Comment (l-wang) [inline src/backend/commands/analyze.c:90]: Should this be part of the second commit?

File: src/backend/commands/analyze.c [hunk 3]
```diff
@@ -1076,8 +1078,8 @@ acquire_sample_rows(Relation onerel, int elevel,
 			 */
 			if (numrows < targrows)
 			{
-				rows[numrows] = ExecCopySlotHeapTuple(slot);
-				rows[numrows]->t_self = slot->tts_tid;
```
Comment (l-wang) [inline src/backend/commands/analyze.c:1080]: Neither `heapam_scan_analyze_sample_tuple` nor `StoreSampleRow()` updates rows[numrows]->t_self, do we need to do this somewhere?

File: src/backend/commands/analyze.c [hunk 4]
```diff
@@ -1134,13 +1134,13 @@ acquire_sample_rows(Relation onerel, int elevel,
 	 */
 	if (bs.m > 0)
 	{
-		*totalrows = floor((liverows / bs.m) * totalblocks + 0.5);
-		*totaldeadrows = floor((deadrows / bs.m) * totalblocks + 0.5);
+		context->totalrows += floor((context->liverows / bs.m) * totalblocks + 0.5);
```
Comment (l-wang) [inline src/backend/commands/analyze.c]: Are `+=` and `=` both correct here?

File: src/include/access/tableam.h [hunk 5]
```diff
@@ -38,6 +38,29 @@ struct TBMIterateResult;
...
+
+	double	totalrows;
+	double 	totaldeadrows;
+	double 	liverows;
+	double 	deadrows;
+	int		numrows; /* number of rows stored in rows array */
+
+#define DISK_DATA 0
+#define DISK_SIZE 1
+#define MAXSAMPLETYPE 2
```
Comment (l-wang) [inline src/include/access/tableam.h]: Might be helpful adding some comments for these macros and the following fields, especially what are expected in slots[DISK_SIZE] and rows[DISK_SIZE].

File: src/include/commands/vacuum.h [hunk 6]
```diff
@@ -113,6 +113,12 @@ typedef struct VacAttrStats
 	int			numvalues[STATISTIC_NUM_SLOTS];
 	Datum	   *stavalues[STATISTIC_NUM_SLOTS];
 
+	/*
+	 * These fields are to be filled in compute_disk_stats
+	 */
+	float4		stadiskfrac;	/* fraction of the physical size */
+	float8		disksize;	/* fraction of the physical size */
```
Comment (l-wang) [inline src/include/commands/vacuum.h]: /* value of the physical size? */

File: src/backend/access/zedstore/zedstore_attstream.c [hunk 7]
```diff
@@ -166,13 +166,15 @@ decode_attstream_begin(attstream_decoder *decoder, ZSAttStream *attstream)
 					  attstream->t_size - SizeOfZSAttStreamHeader,
 					  attstream->t_decompressed_bufsize);
 		decoder->chunks_len = attstream->t_decompressed_size;
+		decoder->compression_ratio = ((float8) buf_size_needed) / attstream->t_size;
```
Comment (l-wang) [inline src/backend/access/zedstore/zedstore_attstream.c:170]: The compression ratio here is only the ratio for zedstore's compression, in-line toasted compression is not considered here, which is probably fine for our purposes. I'm not sure about out-of-line toast pages though, do we need to included the cost of reading the toast page that a toast datum points to?

File: src/backend/access/zedstore/zedstoream_handler.c [hunk 8]
```diff
@@ -2424,34 +2425,109 @@ zedstoream_relation_copy_for_cluster(Relation OldHeap, Relation NewHeap,
...
+	/* initialize the slot to fetch sample rows */
+	InitSampleRows(context, onerel, DISK_DATA);
+
+	/* zedstore also provide extra info when analyzing full columns */
+	if (!va_cols)
+		InitSampleRows(context, onerel, DISK_SIZE);
+
+	/* zedstore use a logical block number to acquire sample rows */
+	tid = zsbt_get_last_tid(onerel);
+	context->totalblocks = ZSTidGetBlockNumber(tid) + 1;
```
Comment (l-wang) [inline src/backend/access/zedstore/zedstoream_handler.c:2454]: Very smart using the logical block number!

**Review by l-wang — APPROVED** [review]
Adding the `stadiskfrac` to pg_statistic, along with the other API changes make sense to me. I have one question about how we calculate the compression ratio for zedstore in my comments, but I'm not too concerned about that. We can merge this to zedstore and propose the API changes to upstream.

File: src/backend/commands/analyze.c (hunk 2 above)
Comment (pengzhout) [reply src/backend/commands/analyze.c:90]: True

File: src/backend/commands/analyze.c (hunk 3 above)
Comment (pengzhout) [reply src/backend/commands/analyze.c:1080]: Nice catch

File: src/backend/commands/analyze.c (hunk 4 above)
Comment (pengzhout) [reply src/backend/commands/analyze.c]: += is used for inherit table, all tables use the same context

File: src/backend/access/zedstore/zedstore_attstream.c (hunk 7 above)
Comment (pengzhout) [reply src/backend/access/zedstore/zedstore_attstream.c:170]: Yes, that's a problem, we have extra blocks in zedstore files include TID blocks, toast blocks and redo blocks, we need a way to consider those blocks. Do heap table considered the cost of toast blocks?

**Comment by pengzhout** [comment]
I didnot use scanCols in this PR for now, we can do that later when scanCols is introduced.

**Pull request merged**